website/blog/2023-03-01.ferretdb-compare-arrays.md

```diff
@@ -0,0 +1,532 @@
+---
+title: "How MongoDB compares arrays in filtering"
+slug: ferretdb-compare-arrays
+authors: [chi]
+description: "In this article, we explore in detail how MongoDB BSON arrays are compared
+in filter operators."
+keywords:
+  - MongoDB $eq operator
+  - MongoDB implicit match
+  - MongoDB $lt operator
+  - MongoDB $gt operator
+  - MongoDB array element match
+  - MongoDB array value match
+image: /img/blog/ferretdb-how-mongodb-compares-arrays.jpg
+unlisted: true
+---
+
+In this article, we explore in detail how MongoDB BSON arrays are compared in filter operators.
+
+![How MongoDB compares arrays](/img/blog/ferretdb-how-mongodb-compares-arrays.jpg)
+
+<!--truncate-->
+
+## Understanding BSON Comparison for Filtering
+
+The purpose of comparing BSON values is to determine which value is
+equal, greater or less than the other.
+The comparison is used in various operations, such as sorting, filter operators,
+as well as update operators.
+
+In this blog post, we will discuss the comparison of BSON values used in filter operators,
+with a focus on arrays.
+We will provide examples to help clarify the process.
+
+### Terminology
+
+**Smallest element** - The smallest element of an array that has the same BSON type
+as the filter argument.
+
+**Largest element** - The largest element of an array that has the same BSON type
+as the filter argument.
+
+**Array values** - A field of a document that holds an array.
+
+### BSON Comparison Order
+
+BSON values have a defined order of comparison, listed from lowest to highest as follows:
+
+1. Null
+2. Numbers
+3. String
+4. Object
+5. Array
+6. BinData
+7. ObjectId
+8. Boolean
+9. Date
+10. Timestamp
+11. Regular Expression
+
+All number types go through conversion, so they have the same BSON comparison order, regardless of
+`double`, `int32`, `int64` or `decimal`.
+
+The BSON comparison order is used for comparing different BSON types.
+If BSON comparison order are equal, we compare their values, if not we do not match it.
+A BSON array may use an element from an array to compare with other BSON type.
+
```

Comment:
I think that's where we should start with examples, starting with basic `$lt` filtering by different numeric types